teps/0099-parameters-in-script.md

```diff
@@ -0,0 +1,479 @@
+
+   Replace direct interpolation with an alternative that does not put the `Parameter`
+   into an executable context by default.
+
+  1. `Tasks` in the *Tekton Catalog* should be bumped to use the safer approach.
+    - Given that this update would likely be backwards-incompatible, as it was in the
+      [`git-clone` Task update in the Catalog][catalog-pr-711], the versions of those
+      `Tasks` have to be bumped with this update.
+    - Catalog `Tasks` that leverage this pattern to intentionally inject executable code
+      (e.g. the `git-cli` Task [accepts a script to run as a param][git-cli]) can be updated
+      to execute the contents of the `Parameter` from a file or call `eval` on it explicitly.
+
+  1. Tekton Pipelines and Catalog documentation should be updated to the new approach.
+
+1. **`Tasks` should not support implicit runtime code injection**.
+
+   Disallow interpolating `Parameter` variables directly into the body of a `Script`
+   by *Tekton Pipelines V1* release to be secure by default.
+
+
+### Use Case
+
+As described in the [original proposal][pipeline-issue-781], the use case for `Script`
+in `Tasks` is:
+
+> Users need an easy cruft-free way to express a multi-statement `Script` in the
+> body of a `Task` `Step`, without having to understand advanced topics about
+> `Containers`, like what an `entrypoint` or a `command` is.
+>
+> Users should only need to have passing familiarity with a shell environment
+> to be successful using Tekton.
+
+Problems with variable subsitution was noted in [original proposal][pipeline-issue-781]:
+
+> How this syntax works with input substitutions, e.g., ``${inputs.params.flags}`` might
+> also lead to confusion among users, which we should design for.
+
+Users need a way to safely and reliably substitute `Parameters` in `Scripts` in `Steps`
+of their `Tasks`. This would make it easier to use Tekton without learning advanced topics,
+while still using it reliably and safely.
+
+### Requirements
+
+1. **[Secure by default][secure-by-default]**: Out of the box
+   *Tekton Pipelines* should disallow common anti-patterns that
+   can render a default install insecure.
+1. **Easy to use**: Whatever replaces the direct use of `$(params.X)` in
+   `script` should be unsurprising and memorable to ease friction in user
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
+1. **Backwards compatible in v1beta1**. If a user already has `Tasks`
+   that use `$(params.X)` syntax in `script` then we should continue supporting
+   that usage in v1beta1 resources. Ideally, Tekton Pipelines would warn the user
+   about it somehow.
+1. **Backwards incompatible in v1**. Starting with v1, Tekton should strive to
```

Comment:
nit: I would rephrase this slightly-- the goal isn't to be backwards incompatible in v1, it's to be secure by default and backwards incompatibility is unfortunately necessary

Reply:
You're totally right @lbernick :  being backwards incompatible is not the end goal here. We'll update this bit.

@vdemeester Maybe we could define backwards compatibility more clearly here?

The TEP is treating "Backwards compatible" to mean we continue to support users injecting executable statements from params into scripts using the current param syntax. E.g. continue to [support the code injection use-case in the `git-clone` catalog task that spurred some of this discussion](https://github.com/tektoncd/catalog/pull/711#issuecomment-845228173).

I'm not sure we can have it both ways though. Today we support the direct interpolation of param values into script bodies as executable statements. Changing that is de facto backwards incompatible I think. As soon as we treat param values as something to sanitize then it necessarily means those values will be escaped before insertion (rendered non-executable), stringified as env vars, put in a file on disk, or something else.